File: 2017/Day15.js
module.exports =function()
{
    function *generatorA1(count, mask) 
    {
        let A = 722;
        while (count > 0) 
        {
            A = (A * 16807) % 2147483647;

            if (mask === undefined || (A & mask) === 0)
            {
                count--;
                yield A & 0xFFFF;
            }
        }
    }

    function *generatorB1(count, mask) 
    {
        let B = 354;
        while (count > 0) 
        {
            B = (B * 48271) % 2147483647;

            if (mask === undefined || (B & mask) === 0)
            {
                count--;
                yield B & 0xFFFF;
            }
        }
    }
      
    function solve(count, maskA, maskB)
    {
        let a = generatorA1(count, maskA);
        let b = generatorB1(count, maskB);
        let matches = 0;

        while(true)
        {
            let v1 = a.next();
            let v2 = b.next();

            if (v1.done || v2.done)
                break;

            if (v1.value === v2.value)
                matches++;
        }

        return matches;
    }

    console.log("Part 1 = " + solve(40000000) + " and Part 2 = " + solve(5000000, 3, 7));
    process.exit(0);    
}

File: 2017/C/Day15.c
#include <stdio.h>
#include <time.h>
#include <sys/types.h>
#include <unistd.h>
#include <sys/wait.h>

static long divider = 2147483647;

static long modulo(long value)
{
	long v = (value & divider) + (value >> 31);
	return (v >= divider) ? v - divider : v;
//	return value % divider;
}

static int solve1()
{
	int matches = 0;
	unsigned long A = 722;
	unsigned long B = 354;
	int count = 40000000;
	
	while (count-- > 0)
	{
		A = modulo(A * 16807);// % 2147483647;
		B = modulo(B * 48271);// % 2147483647;
		
		if ((A & 0xFFFF) == (B & 0xFFFF))
			matches++;
	}
	
	return matches;
}

static int solve2()
{
	int matches = 0;
	unsigned long A = 722;
	unsigned long B = 354;
	int count = 5000000;
	
	while (count-- > 0)
	{
		do
		{
			A = modulo(A * 16807);// % 2147483647;
		}
		while ((A & 3) != 0);
		
		do
		{
			B = modulo(B * 48271);// % 2147483647;
		}
		while ((B & 7) != 0);
		
		if ((A & 0xFFFF) == (B & 0xFFFF))
			matches++;
	}
	
	return matches;
}

void day15(void)
{
	pid_t pid = fork();
	
	if (pid > 0) // The parent
	{
		printf("Part 1: %i\n", solve1());
		
		int status;
		waitpid(pid, &status, 0);
	}
	else if (pid == 0) // the child
	{
		printf("Part 2: %i\n", solve2());
		exit(0);
	}
	else
		printf("Fork failed!\n");
}

